Framework/Algorithms/src/PDCalibration.cpp

```diff
@@ -0,0 +1,808 @@
+    }
+    this->detid = *(detIds.begin());
+
+    const MantidVec &X = spectrum.readX();
+    const MantidVec &Y = spectrum.readY();
+    tofMin = X.front();
+    tofMax = X.back();
+
+    // determine tof min supported by the workspace
+    size_t minIndex = 0; // want to store value
+    for (; minIndex < Y.size(); ++minIndex) {
+      if (isNonZero(Y[minIndex])) {
+        tofMin = X[minIndex];
+        break;
+      }
+    }
+
+    // determin tof max supported by the workspace
+    size_t maxIndex = Y.size() - 1;
+    for (; maxIndex > minIndex; --maxIndex) {
+      if (isNonZero(Y[maxIndex])) {
+        tofMax = X[maxIndex];
+        break;
+      }
+    }
+  }
+
+  void setPositions(const std::vector<double> &peaksInD,
+                    const std::vector<double> &peaksInDWindows,
+                    std::function<double(double)> toTof) {
+
+    const std::size_t numOrig = peaksInD.size();
+    for (std::size_t i = 0; i < numOrig; ++i) {
+      const double centre = toTof(peaksInD[i]);
+      if (centre < tofMax && centre > tofMin) {
+        inDPos.push_back(peaksInD[i]);
+        inTofPos.push_back(peaksInD[i]);
+        inTofWindows.push_back(peaksInDWindows[2 * i]);
+        inTofWindows.push_back(peaksInDWindows[2 * i + 1]);
+      }
+    }
+    std::transform(inTofPos.begin(), inTofPos.end(), inTofPos.begin(), toTof);
+    std::transform(inTofWindows.begin(), inTofWindows.end(),
+                   inTofWindows.begin(), toTof);
+  }
+
+  std::size_t wkspIndex;
+  detid_t detid;
+  double tofMin;
+  double tofMax;
+  std::vector<double> inTofPos;
+  std::vector<double> inTofWindows;
+  std::vector<double> inDPos;
+};
+
+//----------------------------------------------------------------------------------------------
+/** Constructor
+ */
+PDCalibration::PDCalibration() {}
+
+//----------------------------------------------------------------------------------------------
+/** Destructor
+ */
+PDCalibration::~PDCalibration() {}
+
+//----------------------------------------------------------------------------------------------
+
+/// Algorithms name for identification. @see Algorithm::name
+const std::string PDCalibration::name() const { return "PDCalibration"; }
+
+/// Algorithm's version for identification. @see Algorithm::version
+int PDCalibration::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string PDCalibration::category() const {
+  return "Diffraction\\Calibration";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string PDCalibration::summary() const {
+  return "Calibrate the detector pixels and create a calibration table";
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void PDCalibration::init() {
+  declareProperty(Kernel::make_unique<WorkspaceProperty<MatrixWorkspace>>(
```

Comment:
I changed UncalibratedWorkspace to SignalWorkspace as this seems to make more sense with the SignalFile. And added the group.